crates/valence_boss_bar/src/lib.rs

```diff
@@ -1,34 +1,191 @@
+            client.write_packet(&BossBarS2c {
+                id: id.0,
+                action: BossBarAction::UpdateFlags(*flags),
+            });
+        }
+    }
+}
+
+/// System that sends a bossbar add/remove packet to all viewers of a boss bar
+/// that just have been added/removed.
+fn handle_boss_bar_viewers_update(
+    mut boss_bars: Query<
+        (
+            &UniqueId,
+            &BossBarTitle,
+            &BossBarHealth,
+            &BossBarStyle,
+            &BossBarFlags,
+            &mut BossBarViewers,
+        ),
+        Changed<BossBarViewers>,
+    >,
+    mut clients: Query<&mut Client>,
+) {
+    for (id, title, health, style, flags, mut viewers) in boss_bars.iter_mut() {
+        let previous_viewers = &viewers.last_viewers;
+        let current_viewers = &viewers.current_viewers;
+
+        let mut added_viewers = Vec::new();
+        let mut removed_viewers = Vec::new();
+
+        for viewer in viewers.current_viewers.iter() {
+            if !previous_viewers.contains(viewer) {
+                added_viewers.push(*viewer);
+            }
+        }
+
+        for viewer in previous_viewers.iter() {
+            if !current_viewers.contains(viewer) {
+                removed_viewers.push(*viewer);
+            }
+        }
+
+        for added_viewer in added_viewers {
+            let mut client = clients.get_mut(added_viewer).unwrap();
+            client.write_packet(&BossBarS2c {
+                id: id.0,
+                action: BossBarAction::Add {
+                    title: title.0.clone(),
+                    health: health.0,
+                    color: style.color,
+                    division: style.division,
+                    flags: *flags,
+                },
+            });
+        }
+
+        for removed_viewer in removed_viewers {
+            if let Ok(mut client) = clients.get_mut(removed_viewer) {
+                client.write_packet(&BossBarS2c {
+                    id: id.0,
+                    action: BossBarAction::Remove,
+                });
+            }
+        }
 
+        viewers.last_viewers = viewers.current_viewers.clone();
+    }
 }
 
-fn remove_despawned_boss_bars_from_viewers(mut boss_bars: Query<(&UniqueId, &mut BossBarViewers), Added<Despawned>>, mut clients: Query<&mut Client>) {
+/// System that sends a bossbar remove packet to all viewers of a boss bar that
+/// has been despawned.
+fn handle_boss_bar_despawn(
+    mut boss_bars: Query<(&UniqueId, &mut BossBarViewers), Added<Despawned>>,
+    mut clients: Query<&mut Client>,
+) {
     for boss_bar in boss_bars.iter_mut() {
         let (id, mut viewers) = boss_bar;
 
-        for viewer in viewers.0.iter_mut() {
+        for viewer in viewers.current_viewers.iter_mut() {
             let mut client = clients.get_mut(*viewer).unwrap();
             client.write_packet(&BossBarS2c {
                 id: id.0,
                 action: BossBarAction::Remove,
             });
         }
     }
-}
```

Comment:
You don't remove the bossbar from the world

Reply:
Isn't there a system that automatically removes entities from the world when they get the Despawned component? Otherwise I will do it, thanks.

Reply:
Ho I didn't think of that, You are right it is done automatically
https://github.com/valence-rs/valence/blob/d36998dc9cba25aba4b9287db41ce69911708775/crates/valence_core/src/despawn.rs#L19-L26

Reply:
Yes, but you need to detect when the boss bar gains the `Despawned` component and send the appropriate packets to despawn the bar for all viewers.